### PR TITLE
Add pause/resume control to AudioOut and I2SOut

### LINKED ISSUE #808: pause/unpause support in audioio
had a project we wanted to pause audio wave playback and found we could not - it would be a nice addition :)

### PR DESCRIPTION
**Pull request by tannewt**
Fixes #808

### CODE REVIEW

File: shared-bindings/audiobusio/I2SOut.c [hunk 1]
```diff
@@ -212,16 +212,69 @@ const mp_obj_property_t audiobusio_i2sout_playing_obj = {
...
+//|   .. method:: pause()
+//|
+//|     Stops playback temporarily while remembering the position. Use `resume` to resume playback.
+//|
+STATIC mp_obj_t audiobusio_i2sout_obj_pause(mp_obj_t self_in) {
+    audiobusio_i2sout_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    raise_error_if_deinited(common_hal_audiobusio_i2sout_deinited(self));
+
+    if (!common_hal_audiobusio_i2sout_get_playing(self)) {
+        mp_raise_RuntimeError("No sample playing cannot pause");
```
Comment (dhalbert) [inline shared-bindings/audiobusio/I2SOut.c]: I was going to suggest adding a ":" to make "No sample playing: cannot pause", but how about just "Not playing"?

File: shared-bindings/audioio/AudioOut.c [hunk 2]
```diff
@@ -209,16 +209,69 @@ const mp_obj_property_t audioio_audioout_playing_obj = {
...
+//|   .. method:: pause()
+//|
+//|     Stops playback temporarily while remembering the position. Use `resume` to resume playback.
+//|
+STATIC mp_obj_t audioio_audioout_obj_pause(mp_obj_t self_in) {
+    audioio_audioout_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    raise_error_if_deinited(common_hal_audioio_audioout_deinited(self));
+
+    if (!common_hal_audioio_audioout_get_playing(self)) {
+        mp_raise_RuntimeError("No sample playing cannot pause");
```
Comment (dhalbert) [inline shared-bindings/audioio/AudioOut.c]: I was going to suggest adding a ":" to make "No sample playing: cannot pause", but how about just "Not playing"?

**Comment by dhalbert** [comment]
Also if a sample has never started playing, could `resume()` be the equivalent of `play()`?

**Comment by deshipu** [comment]
Or even better, only have `play()`.

**Comment by dhalbert** [comment]
If you just have `play()`, then you need some way to rewind or seek to restart playing form the beginning.

File: shared-bindings/audiobusio/I2SOut.c (hunk 1 above)
Comment (tannewt) [reply shared-bindings/audiobusio/I2SOut.c]: Done

File: shared-bindings/audioio/AudioOut.c (hunk 2 above)
Comment (tannewt) [reply shared-bindings/audioio/AudioOut.c]: Done

**Comment by dhalbert** [comment]
maybe play-stop does a rewind. play-pause-play is like resume. This is the way music player buttons work, right?

**Comment by tannewt** [comment]
The new API also has the sample itself provided to `play(sample)`. (Unlike 2.x which had the sample provided in the constructor.) I like `resume` separate because it doesn't require the sample again and makes it clear that its simply resuming whatever was already going.

**Comment by dhalbert** [comment]
do you want to regularize pause-pause vs resume-resume?

**Comment by tannewt** [comment]
Yup! Just saw that. Added in the latest commit.

**Pull request merged**